File: abstractions/java/lib/src/main/java/com/microsoft/kiota/serialization/ParseNodeFactoryRegistry.java
package com.microsoft.kiota.serialization;

import java.io.InputStream;
import java.util.HashMap;
import java.util.Objects;
import java.util.regex.Pattern;

import javax.annotation.Nonnull;

/**
 * This factory holds a list of all the registered factories for the various types of nodes.
 */
public class ParseNodeFactoryRegistry implements ParseNodeFactory {
    /** Default singleton instance of the registry to be used when registering new factories that should be available by default. */
    public static final ParseNodeFactoryRegistry defaultInstance = new ParseNodeFactoryRegistry();
    /** List of factories that are registered by content type. */
    public HashMap<String, ParseNodeFactory> contentTypeAssociatedFactories = new HashMap<>();
    public String getValidContentType() {
        throw new UnsupportedOperationException("The registry supports multiple content types. Get the registered factory instead.");
    }
    private static Pattern contentTypeVendorCleanupPattern = Pattern.compile("[^/]+\\+", Pattern.CASE_INSENSITIVE);
    @Override
    @Nonnull
    public ParseNode getParseNode(@Nonnull final String contentType, @Nonnull final InputStream rawResponse) {
        Objects.requireNonNull(contentType, "parameter contentType cannot be null");
        Objects.requireNonNull(rawResponse, "parameter rawResponse cannot be null");
        if(contentType.isEmpty()) {
            throw new NullPointerException("contentType cannot be empty");
        }
        final String vendorSpecificContentType = contentType.split(";")[0];
        if(contentTypeAssociatedFactories.containsKey(vendorSpecificContentType)) {
            return contentTypeAssociatedFactories.get(vendorSpecificContentType).getParseNode(vendorSpecificContentType, rawResponse);
        }
        final String cleanedContentType = contentTypeVendorCleanupPattern.matcher(vendorSpecificContentType).replaceAll("");
        if(contentTypeAssociatedFactories.containsKey(cleanedContentType)) {
            return contentTypeAssociatedFactories.get(cleanedContentType).getParseNode(cleanedContentType, rawResponse);
        }
        throw new RuntimeException("Content type " + cleanedContentType + " does not have a factory to be parsed");
    }
}

File: abstractions/python/kiota/abstractions/serialization/parse_node_factory_registry.py
from __future__ import annotations

import re
from io import BytesIO
from typing import Dict

from .parse_node import ParseNode
from .parse_node_factory import ParseNodeFactory


class ParseNodeFactoryRegistry(ParseNodeFactory):
    """Holds a list of all the registered factories for the various types of nodes
    """
    CONTENT_TYPE_ASSOCIATED_FACTORIES: Dict[str, ParseNodeFactory] = {}

    __instance = None

    def __new__(cls, *args, **kwargs):
        """Default singleton instance of the registry to be used when registering new
        factories that should be available by default.

        Returns:
            [ParseNodeFactoryRegistry]: Default singleton instance of the class
        """
        if not ParseNodeFactoryRegistry.__instance:
            ParseNodeFactoryRegistry.__instance = object.__new__(cls)
        return ParseNodeFactoryRegistry.__instance

    def get_valid_content_type(self) -> str:
        raise Exception(
            "The registry supports multiple content types. Get the registered factory instead"
        )

    def get_root_parse_node(self, content_type: str, content: BytesIO) -> ParseNode:
        if not content_type:
            raise Exception("Content type cannot be null")
        if not content:
            raise Exception("Content cannot be null")

        vendor_specific_content_type = content_type.split(';')[0]
        factory = self.CONTENT_TYPE_ASSOCIATED_FACTORIES.get(vendor_specific_content_type)
        if factory:
            return factory.get_root_parse_node(vendor_specific_content_type, content)

        cleaned_content_type = re.sub(r'[^/]+\+', '', vendor_specific_content_type)
        factory = self.CONTENT_TYPE_ASSOCIATED_FACTORIES.get(cleaned_content_type)
        if factory:
            return factory.get_root_parse_node(cleaned_content_type, content)

        raise Exception(
            f"Content type {cleaned_content_type} does not have a factory registered to be parsed"
        )
